### DISCUSSION TITLE
Dynamically load timezones list from a resource route - how to?

### DISCUSSION THREAD

**Discussion by meglio** [remix-run/remix#4314] [Q&A] [answered]
Imagine I have a TimezoneSelector react component. I would like to only load the huge list of all timezones when the user clicks on the timezone selector. The spinner would be shown for a moment (while the list of timezones is being fetched - i.e. lazily), and then the list is shown.

With using the standard tools that Remix provides, what would be a good/canonical way to fetch timezones dynamically (i.e. by request, i.e. lazily)?

**Comment by kiliman** [comment]
You should use the `useFetcher` hook. Something like this:

* `fetcher.load()` calls your API
* `fetcher.state` is used to control spinner
* `fetcher.data` contains the data from your API

```ts
function TimezoneSelector() {
  const fetcher = useFetcher()
  const handleClick = () => {
    fetcher.load('/api/timezones')
  }
  return (
    <div>
      <button onClick={handleClick}>Get Timezones</button>
      { fetcher.state === 'submitting' && <Spinner /> }
      <TimezoneList items={fetcher.data} />
    </div>
  )
}
```
https://remix.run/docs/en/v1/api/remix#usefetcher

**Reply by meglio** [reply]
Thanks for the suggerstion! Now, if I have multiple components on the page, and only want this list to be loaded once only and shared between all of the components, what would be the canonical way of doing it the Remix-way?

**Reply by kiliman** [reply]
You can return the timezone data from your route loader. Components that are on the route can call `useLoaderData` to get the data. 

Unfortunately loaders must be exported from the route. Remix doesn't support component-level loaders.

I have a POC that shows how to add loaders/actions from your component (NOTE: You still have to call it from the route, though)

https://github.com/kiliman/remix-component-data

**Comment by meglio** [comment, marked as answer]
Solved in the following way:

1. Encapsulate timezones in a `zustand` global store.
2. Initialize the store with the fetcher in `useEffect` in `root.jsx`.
3. In the timezone dropdown component, ask the global store for the timezones list, and that is why they get loaded via the Remix-provided fetcher lazily.
4. Because the store is global, the timezones list is shared between all components throughout the application. In the same time, the list is only loaded upon first request.

Here is an example of such a store. It misses function definitions that are out of scope of this discussion, but you get the idea:

In `timezones_store.js`:

```jsx

const NEVER_LOADED_PARTIAL_STATE = {
    timezones: undefined,
    timezone_name_to_item: {},
    num_load_calls: 0,
    filter_items: (items, input_value) => { return items; },
    loader_call_scheduled: false,
    error: undefined,
};

export const use_timezones_store = create((set, get) => {

    const load = () => {
        const {
            num_load_calls,
            loader_function,
        } = get();

        // The loader function was never called before. Call it now or schedule until when it is set.
        if (num_load_calls === 0) {
            if (loader_function) {
                // Call the loader function.
                // console.log('Timezones Store: calling the loader_function().');
                set({num_load_calls: 1});
                loader_function();
            } else {
                // console.log('Timezones Store: scheduling call to the loader_function() because it is not set yet.');
                set({num_load_calls: 1, loader_call_scheduled: true});
            }

        } else {
            // Loader was called already, ignore.
            // console.log('Ignoring call to the loader_function() because it was called already.');
        }
    };

    const set_loader_function = (loader_function) => {
        // console.log('Timezones Store: setting the loader_function().');
        const {loader_call_scheduled} = get();
        if (loader_call_scheduled) {
            // console.log(`Timezones Store: calling the loader_function() because it was scheduled.`);
            set({loader_function: loader_function, loader_call_scheduled: false});
            loader_function();
        } else {
            set({loader_function: loader_function});
        }

    }

    const set_timezones = (timezones) => {
        const timezones_meta = get_timezones_meta(timezones);

        // console.log('Timezones Store: set_timezones(), meta = ', timezones_meta);

        set({timezones, ...timezones_meta});
    }

    return {
        ... NEVER_LOADED_PARTIAL_STATE,
        set_loader_function,
        set_timezones,
        load,
    }
});
```

In `root.jsx`:

```js
// Timezones Fetcher.
    const timezones_fetcher = useFetcher();
    const {
        set_loader_function: timezones_store_set_loader_function,
        set_timezones,
    } = use_timezones_store();
    useEffect(() => {
        const loader_function = () => {
            timezones_fetcher.load("/act/ui/timezones");
        }
        timezones_store_set_loader_function(loader_function);
    }, [timezones_store_set_loader_function, timezones_fetcher.load]);
    useEffect(() => {
        const timezones = timezones_fetcher?.data?.data?.timezones;
        if (timezones) {
            set_timezones(timezones);
        }
    }, [timezones_fetcher.data]);
```

It all looks ugly and over-enginnered overall, but it does exactly what I need while using Remix' native fetcher, so is a good template until Remix comes with a better solution.
